presto-main/src/main/java/com/facebook/presto/operator/project/PageProcessor.java

```diff
@@ -15,11 +15,153 @@
 
 import com.facebook.presto.spi.ConnectorSession;
 import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.block.Block;
+import com.google.common.collect.AbstractIterator;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterators;
 
-public interface PageProcessor
+import javax.annotation.concurrent.NotThreadSafe;
+
+import java.util.List;
+import java.util.Optional;
+
+import static com.facebook.presto.operator.project.PageProcessorOutput.EMPTY_PAGE_PROCESSOR_OUTPUT;
+import static com.google.common.base.Verify.verify;
+import static java.util.Objects.requireNonNull;
+
+@NotThreadSafe
+public class PageProcessor
 {
-    /**
-     * @return filtered and projected page, null if all positions are filtered
-     */
-    PageProcessorOutput process(ConnectorSession session, Page page);
+    static final int MAX_BATCH_SIZE = 8 * 1024;
+    static final int MAX_PAGE_SIZE_IN_BYTES = 4 * 1024 * 1024;
```

Comment:
This code is pretty different from traditional page building because it is columnar.  The heuristic works best when there is a big enough range for doubling and halving to not always be jumping around.  This code also has a row limit to prevent having 1 MB of boolean values.

I think we should make similar changes to `PageBuilder` to limit max number of rows in a page and then maybe increase the page size, but I don't want to put that is in this PR.